pkg/service/service.go

```diff
@@ -1684,13 +1732,20 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []*lb.Backend
 
 	for hash, backend := range svc.backendByHash {
 		if _, found := backendSet[hash]; !found {
-			obsoleteSVCBackendIDs = append(obsoleteSVCBackendIDs, backend.ID)
-			if s.backendRefCount.Delete(hash) {
-				DeleteBackendID(backend.ID)
-				delete(s.backendByHash, hash)
-				obsoleteBackendIDs = append(obsoleteBackendIDs, backend.ID)
+			if svc.restoredBackendHashes.Has(hash) {
+				// Don't treat backends restored from the datapath and not yet observed as
+				// obsolete, because that would cause connections targeting those backends
+				// to be dropped in case we haven't fully synchronized yet.
+				backends = append(backends, backend)
+			} else {
+				obsoleteSVCBackendIDs = append(obsoleteSVCBackendIDs, backend.ID)
+				if s.backendRefCount.Delete(hash) {
+					DeleteBackendID(backend.ID)
+					delete(s.backendByHash, hash)
+					obsoleteBackendIDs = append(obsoleteBackendIDs, backend.ID)
+				}
+				delete(svc.backendByHash, hash)
```

Comment:
Since there is no change in the original code to delete obsolete backends, can we wrap the special restore case inside `if` block and break?